docs/upgrade.rst

```diff
@@ -0,0 +1,317 @@
+  ``py::module_local(false)``.
+
+The latter is an easy workaround, but in the long run it would be best to
+localize all common type bindings in order to avoid conflicts with
+third-party modules.
+
+
+New syntax for custom constructors
+----------------------------------
+
+The old placement-new custom constructors are still valid, but the new approach
+greatly improves type safety (placement-new can be called accidentally with an
+incompatible type, without any compiler errors or warnings). See
+:ref:`custom_constructors` for details.
+
+.. code-block:: cpp
+
+    // old
+    py::class<Foo>(m, "Foo")
+        .def("__init__", [](Foo &self, ...) {
+            new (&self) Foo(...); // uses placement-new
+        });
+
+    // new
+    py::class<Foo>(m, "Foo")
+        .def(py::init([](...) {
+            return new Foo(...); // return by raw pointer
+            // or: return std::make_unique<Foo>(...); // return by holder
+            // or: return Foo(...); // return by value (move constructor)
+        }));
+
+
+Deprecation of some ``py::object`` APIs
+---------------------------------------
+
+The equality comparison operator for ``py::object`` only compares the
+underlying pointers, therefore ``obj1 == obj2`` in C++ is equivalent to the
+expression ``obj1 is obj2`` in Python. To avoid confusion, the equality
+operator now emits a deprecation warning (in the future, it can be replaced
+with proper rich object comparison). To compare objects by pointer, use
+``obj1.is(obj2)``.
+
+For classes which inherit from ``py::object``, ``borrowed`` and ``stolen``
+were previously available as protected constructor tags. Now the types
+should be used directly instead: ``borrowed_t{}`` and ``stolen_t{}``
+(`#771 <https://github.com/pybind/pybind11/pull/771>`_).
+
+
+v2.1
+====
+
+Minimum compiler versions are enforced at compile time
+------------------------------------------------------
+
+The minimums also apply to v2.0 but the check is now explicit and a compile-time
+error is raised if the compiler does not meet the requirements:
+
+* GCC >= 4.8
+* clang >= 3.3 (appleclang >= 5.0)
+* MSVC >= 2015u3
+* Intel C++ >= 15.0
+
+
+Deprecation of the ``py::metaclass()`` attribute
+------------------------------------------------
+
+It's no longer required for binding classes with static properties.
+
+.. code-block:: cpp
+
+    // old -- emits a deprecation warning
+    py::class_<Foo>(m, "Foo", py::metaclass())
+        .def_property_readonly_static("foo", ...);
+
+    // new -- works without the attribute
+    py::class_<Foo>(m, "Foo")
+        .def_property_readonly_static("foo", ...);
+
+
+v2.0
+====
+
+Breaking changes in ``py::class_``
+----------------------------------
+
+These changes were necessary to make type definitions in pybind11
+future-proof, to support PyPy via its ``cpyext`` mechanism (`#527
+<https://github.com/pybind/pybind11/pull/527>`_), and to improve efficiency
+(`rev. 86d825 <https://github.com/pybind/pybind11/commit/86d825>`_).
```

Comment:
The referenced change (86d825 - moving trampolines declaration into the `class_<...>` template types) looks like it was already in 1.8.  Should that be a different commit?